src/backend/distributed/executor/executor.c

```diff
@@ -1346,6 +1407,40 @@ ConnectionStateMachine(WorkerSession *session)
 }
 
 
+/*
+ * Use2PCIfParticipatingNodesExpanded ensures that the execution uses
+ * 2PC if the transaction is expanded to access a new worker node.
+ */
+static void
+Use2PCIfParticipatingNodesExpanded(WorkerSession *session)
+{
+	DistributedExecution *execution = NULL;
+
+	if (DoesCoordinatedTransactionUse2PC() ||
+		MultiShardCommitProtocol != COMMIT_PROTOCOL_2PC)
+	{
+		/* already using or doesn't require 2PC */
+		return;
```

Comment:
Given how fast the operation below is, I don't think it's really worth having this code block and the extra function.